File: game/cave/lost.py
import pygame

from animation import Animation
from cave.cave_resources import CaveResources
from cave.lost_spring import LostSpring
from game_data import GameData
from phone_events import PhoneEvents
from state import State


class Lost(State):
    def __init__(self, context: GameData):
        super().__init__(context)
        self.animation = [
            CaveResources.hugo_puff_first,
            CaveResources.hugo_puff_second,
            CaveResources.hugo_puff_third
        ][context.cave_selected_rope]

    def process_events(self, phone_events: PhoneEvents):
        if self.get_frame_index() >= len(self.animation):
            return LostSpring

        if self.one_shot(1, "PrePuf"):
            pygame.mixer.Sound.play(CaveResources.pre_puf)

        if self.one_shot(2, "Puf"):
            pygame.mixer.Sound.play(CaveResources.puf)

        return None

    def render(self, screen):
        screen.blit(Animation.get_frame(self.animation, self.get_frame_index()), (0, 0))

File: game-web/src/cave/lost.ts
/** Port of game/cave/lost.py */

import { State, type StateClass } from '../state';
import { Animation } from '../animation';
import { AudioHelper } from '../audioHelper';
import { CaveResources } from './caveResources';
import { LostSpring } from './lostSpring';
import type { GameData } from '../gameData';
import type { PhoneEvents } from '../phoneEvents';

export class Lost extends State {
  private animation: ImageBitmap[];

  constructor(context: GameData) {
    super(context);
    this.animation = [
      CaveResources.hugoPuffFirst,
      CaveResources.hugoPuffSecond,
      CaveResources.hugoPuffThird,
    ][context.caveSelectedRope];
  }

  processEvents(_events: PhoneEvents): StateClass | null {
    if (this.getFrameIndex() >= this.animation.length) return LostSpring;

    if (this.oneShot(1, 'PrePuf')) AudioHelper.play(CaveResources.prePuf);
    if (this.oneShot(2, 'Puf'))    AudioHelper.play(CaveResources.puf);

    return null;
  }

  render(ctx: OffscreenCanvasRenderingContext2D): void {
    ctx.drawImage(Animation.getFrame(this.animation, this.getFrameIndex()), 0, 0);
  }
}
